gear-programs/vft-treasury/src/services/mod.rs

```diff
@@ -0,0 +1,122 @@
+use gstd::{debug, exec, msg};
+use sails_rs::{gstd::ExecContext, prelude::*};
+
+mod utils;
+mod vft;
+
+pub struct VftTreasury<ExecContext> {
+    exec_context: ExecContext,
+}
+
+static mut DATA: Option<VftTreasuryData> = None;
+
+#[derive(Debug, Default)]
+struct VftTreasuryData {
+    admin: ActorId,
+    ethereum_event_client: ActorId,
+}
+
+#[derive(Debug, Decode, Encode, TypeInfo)]
+pub struct InitConfig {
+    pub ethereum_event_client: ActorId,
+}
+
+impl InitConfig {
+    pub fn new(ethereum_event_client: ActorId) -> Self {
+        Self {
+            ethereum_event_client,
+        }
+    }
+}
+impl<T> VftTreasury<T>
+where
+    T: ExecContext,
+{
+    pub fn seed(config: InitConfig, exec_context: T) {
+        unsafe {
+            DATA = Some(VftTreasuryData {
+                ethereum_event_client: config.ethereum_event_client,
+                admin: exec_context.actor_id(),
+            });
+        }
+    }
+
+    pub fn new(exec_context: T) -> Self {
+        Self { exec_context }
+    }
+
+    fn data(&self) -> &VftTreasuryData {
+        unsafe { DATA.as_ref().expect("VftTreasury::seed() must be called") }
+    }
+
+    fn data_mut(&mut self) -> &mut VftTreasuryData {
+        unsafe { DATA.as_mut().expect("VftTreasury::seed() must be called") }
+    }
+}
+
+#[derive(Encode, Decode, TypeInfo)]
+pub enum VftTreasuryEvents {
+    Deposit {
+        from: ActorId,
+        to: H160,
+        token: ActorId,
+        amount: U256,
+    },
+    Withdraw {
+        receiver: ActorId,
+        token: ActorId,
+        amount: U256,
+    },
+}
+
+#[service(events = VftTreasuryEvents)]
+impl<T> VftTreasury<T>
+where
+    T: ExecContext,
+{
+    pub fn update_ethereum_event_client_address(&mut self, new_address: ActorId) {
+        let data = self.data();
+        if data.admin != self.exec_context.actor_id() {
+            panic!("Not admin");
+        }
+
+        self.data_mut().ethereum_event_client = new_address;
+    }
+
+    pub async fn deposit(&mut self, token: ActorId, amount: U256, to: H160) {
```

Comment:
I think we should pass the sender's address here:


    pub async fn deposit(&mut self, sender: ActorId, token: ActorId, amount: U256, to: H160) {

because I won't be transferring tokens to the bridge-payment address.

Reply:
cc @mertwole

Reply:
yes, agree with that

Reply:
Also `deposit` method should probably return mapped `eth_token_id`, 
it will be used in `bridging-payment` `BridgingPaymentEvents` event.

Reply:
Also `nonce` will be used by `relayer` to distinguish sent messages